### PR TITLE
Function to linearly interpolate state mutable sequences

### PR DESCRIPTION
**Pull request by gawebb-dstl**
The function will take a `StateMutableSequence` and a `times: List[datetime.datetime]`. The function will create a new `StateMutableSequence` with states only at the times specified by `times`.

This time with a clean git history. See #865 for messy history

This provides a step for fixing #364

### CODE REVIEW

**Comment by gawebb-dstl** [comment]
I've looked at 4 options (see code). All 4 options run function correctly. Option 4 is the quickest* so will be taken forward

Output when running the file:
```
Option 1 took 3.9378470000000005
Option 2 took 79.7136048
Option 3 took 1.762432000000004
Option 4 took 1.326851700000006
Option 1 worked
Option 2 worked
Option 3 worked
Option 4 worked
```

*Option 3 is quicker if all of the states need to be produced by interpolation

File: stonesoup/functions/interpolate.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        times = new_times
+
+    # Find times that require interpolation
+    times_to_interpolate = sorted(set(times).difference(time_state_dict.keys()))
+
+    if len(times_to_interpolate) > 0:
+        # Only interpolate if required
+        state_vectors = [state.state_vector for state in time_state_dict.values()]
+        state_times = [time.timestamp() for time in time_state_dict.keys()]
+        interpolate_object = interp1d(state_times, np.stack(state_vectors, axis=0), axis=0)
```
[the remark this answers is not in the document]
Comment (gawebb-dstl) [reply stonesoup/functions/interpolate.py]: I've changed it to numpy.interp now

File: docs/source/stonesoup.functions.rst [hunk 2]
```diff
@@ -5,3 +5,5 @@ Functions
 
 .. automodule:: stonesoup.functions.orbital
 
+.. automodule:: stonesoup.functions.interpolate
+
```
Comment (sdhiscocks) [inline docs/source/stonesoup.functions.rst:9]: Both orbital and interpolate modules should probably have a title.

File: stonesoup/functions/interpolate.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    # A StateMutableSequence is produced by the inner function call.
+    # The single state is taken from that StateMutableSequence
+    if isinstance(times, datetime.datetime):
+        new_sms = interpolate_state_mutable_sequence(sms, [times])
+        return new_sms.state
+
+    # Track metadata removed and no interpolation can be performed on the metadata
+    new_sms = copy.copy(sms)
+    if hasattr(new_sms, "metadatas"):
+        new_sms.metadatas = list()
```
Comment (sdhiscocks) [inline stonesoup/functions/interpolate.py:70]: Doc string should mention that `metadata` will be removed. Would be good to resolve this in future.

**Review by sdhiscocks — APPROVED** [review]

**Review by jswright-dstl — APPROVED** [review]

**Pull request merged**